R/secret.R

```diff
@@ -0,0 +1,134 @@
+#   - secret_write() helps you store an encrypted file below `inst/secret` in
+#     the source of 'package':
+#     secret_write(
+#       package = "gargle",
+#       name = "gargle-testing.json",
+#       data = "~/.R/gargle/gargle-testing.json"
+#     )
+#   - We anticipate doing this with the JSON files that hold service account
+#     tokens.
+# * Rig tests (or examples or whatever) to fail gracefully if decryption is not
+#   going to succeed, i.e. if the PASSWORD is not found or the Suggested sodium
+#   package is not installed.
+#   - Here's an example from bigrquery/tests/testthat/setup-auth.R
+#     if (secret_can_decrypt("bigrquery")) {
+#       json <- secret_read("bigrquery", "service-token.json")
+#       bq_auth(path = rawToChar(json))
+#     }
+#   - Then this skipper is defined and used:
+#     skip_if_no_auth <- function() {
+#       testthat::skip_if_not(has_access_cred(), "Authentication not available")
+#     }
+#   - Here's an example with gargle alone:
+#     token <- credentials_service_account(
+#       scopes = "https://www.googleapis.com/auth/userinfo.email",
+#       path = rawToChar(secret_read("gargle", "gargle-testing.json"))
+#     )
+#     get_userinfo(token)
+
+# Setup support for the NAME=PASSWORD envvar ----------------------------------
+
+## secret_pw_name("gargle") --> "GARGLE_PASSWORD"
+secret_pw_name <- function(package) {
+  paste0(toupper(package), "_PASSWORD")
+}
+
+secret_pw_gen <- function() {
+  x <- sample(c(letters, LETTERS, 0:9), 50, replace = TRUE)
+  paste0(x, collapse = "")
+}
+
+secret_pw_try <- function(package) {
+  envvar <- secret_pw_name(package)
+  Sys.getenv(envvar, "")
+}
+
+secret_pw_exists <- function(package) {
+  !identical(secret_pw_try(package), "")
+}
+
+secret_pw_get <- function(package) {
+  env <- secret_pw_name(package)
+  pass <- Sys.getenv(env, "")
+  if (identical(pass, "")) {
+    stop_glue("Envvar {sq(env)} not defined")
+  }
+
+  sodium::sha256(charToRaw(pass))
+}
+
+# Store and retrieve encrypted data -------------------------------------------
+
+secret_can_decrypt <- function(package) {
+  requireNamespace("sodium", quietly = TRUE) && secret_pw_exists(package)
+}
+
+secret_write <- function(package, name, data) {
+  #if (inherits(data, "connection")) {
+    data <- readBin(data, "raw", file.size(data))
+  #} else if (is.character(data)) {
+  #  data <- charToRaw(data)
+  #}
+
+  secret <- fs::path("inst", "secret")
+  if (!fs::dir_exists(secret)) {
+    fs::dir_create(secret)
+  }
+  dst <- fs::path(secret, name)
+
+  enc <- sodium::data_encrypt(
+    data,
+    key = secret_pw_get(package),
+    nonce = secret_nonce()
+  )
+  attr(enc, "nonce") <- NULL
+  writeBin(enc, dst)
+
+  invisible(dst)
+}
+
+# Generated with sodium::bin2hex(sodium::random(24)). AFAICT nonces are
+# primarily used to prevent replay attacks, which shouldn't be a concern here
+secret_nonce <- function() {
+  sodium::hex2bin("cb36bab652dec6ae9b1827c684a7b6d21d2ea31cd9f766ac")
```

Comment:
Generated a new nonce for gargle.

Reply:
I don't _think_ this is necessary but I'm not sure.